python/tvm/relax/transform/mixed_precision.py

```diff
@@ -0,0 +1,135 @@
+DEFAULT_FOLLOW_LIST = [
+    "relax.nn.flatten",
+    "relax.nn.batch_norm",
+    "relax.nn.dropout",
+    "relax.nn.max_pool2d",
+    "relax.ewise_fma",
+    "relax.transpose",
+    "relax.reshape",
+    "relax.expand_dims",
+    "relax.squeeze",
+    "relax.unique",
+    "relax.nn.relu",
+    "relax.nn.gelu",
+    "relax.multiply",
+    "relax.add",
+    "relax.nn.silu",
+    "relax.sqrt",
+    "relax.divide",
+    "relax.subtract",
+    "relax.strided_slice",
+    "relax.sin",
+    "relax.cos",
+    "relax.concatenate",
+    "relax.image.resize2d",
+    "relax.cast",
+    "relax.broadcast_to",
+]
+DEFAULT_NEVER_LIST = ["relax.nn.softmax", "relax.nn.layer_norm", "relax.sum", "relax.mean"]
+
+# Returns a decorator which registers for every given op, the function under FMixedPrecision
+
+
+def register_func_to_op_list(list_ops: List):
+    def decorator(func):
+        for op_name in list_ops:
+            tvm.ir.register_op_attr(op_name, "FMixedPrecision", func, level=10)
+
+    return decorator
+
+
+def infer_mixed_precision(call_node: Call, out_dtype: str) -> List:
+    """A function which returns the infer result of mixed precision for a given call node.
+    Parameters
+    ---------
+    call_node: Call
+        The call node containing the op.
+    out_dtype: str
+        The output data type of gemm/conv, which is the data type of the accumulator.
+    Returns
+    -------
+    has_out_dtype: bool
+        Whether the op has an out_dtype attribute.
+    adjust_call: Call
+        The call node with the out_dtype attribute set to out_dtype.
+    """
+    if hasattr(call_node.attrs, "out_dtype"):
+        if call_node.op.name == "relax.nn.conv2d":
+            adjust_call = conv2d(
+                call_node.args[0],
+                call_node.args[1],
+                strides=call_node.attrs.strides,
+                padding=call_node.attrs.padding,
+                dilation=call_node.attrs.dilation,
+                data_layout=call_node.attrs.data_layout,
+                kernel_layout=call_node.attrs.kernel_layout,
+                out_layout=call_node.attrs.out_layout,
+                out_dtype=out_dtype,
+            )
+        elif call_node.op.name == "relax.nn.matmul":
+            adjust_call = matmul(
+                call_node.args[0],
+                call_node.args[1],
+                out_dtype=out_dtype,
+            )
+        else:
+            raise ValueError("Unsupported op for get_generic_out_dtypes", call_node.op.name)
+        return [True, adjust_call]
+    else:
+        return [False, call_node]
+
+
+# Functions for FMixedPrecision which take in Call and a dtype and returns a MixedTypePolicy,
+# and flag indicating whether the call node has an out_dtype attribute, and the call node with
+# the out_dtype attribute set to out_dtype.
+@register_func_to_op_list(list_ops=DEFAULT_ALWAYS_LIST)
+def generic_always_op(call_node: Call, out_dtype: str) -> List:
```

Comment:
Only specific type of special note need infer_mixed_precision. To simplify the logic, I would have two separate attributes:
- MixedPrecisionPolicyKind which just registers the mixed precision policy
- AdjustMixedPrecison, whch returns an Optional<Call> that contains the adjusted node if it is needed, otherwise. I am not sure has_out_dtype is necessary here

Reply:
Yeah I can remove it.